SEP 22: Move old-releases to a different domain name/url (#28)
saltstack/salt-enhancement-proposals: old-releases.md

- Feature Name: move non-supported releases to a separate domain name
- Start Date: 2020-05-20
- SEP Status: Draft
- SEP PR: (leave this empty)
- Salt Issue: (leave this empty)

# Summary
[summary]: #summary
With the recent CVE events, we have observed and have been alerted to
situations where clients and users were accessing and downloading older
unsupported releases that contain the vulnerability. The ease of access,
coupled with the potential for error that we observed, mean that we need to
make a decision on what to do with older, unsupported releases.

We have three potential paths: release updated packages for older versions,
remove older versions all together, or move older versions to an archive
location that is separated from currently supported versions. From a time and
effort standpoint, the option to update is simply not viable and would set the
project back considerably. The option to completely remove the older packages
would leave many stranded while trying to upgrade to later versions and is
therefore not viable either. This leaves us with the approach of moving the

> older, unsupported version to an archive location. In examining this, we

Reviewer: ```suggestion
older, unsupported versions to an archive location. In examining this, we
```

assessed other open projects and their approach to older, unsupported releases.
We found many following the same proposed approach that we have in this SEP. An

> example is Ubuntu, which makes this distinction easy by using two seperate

Reviewer: ```suggestion
example is Ubuntu, which makes this distinction easy by using two separate
```

have to sync as much to get all supported releases. This reduces bandwidth
costs for them and SaltStack.

# Design
[design]: #detailed-design
Whenever a major release goes out of CVE support, we will move the contents to
https://old-releases.repo.saltstack.com, for example:

```
https://repo.saltstack.com/yum/redhat/7/x86_64/2018.3/
```
would move to

```
https://old-releases.repo.saltstack.com/yum/redhat/7/x86_64/2018.3/
```
and all of the point releases under

```
https://repo.saltstack.com/yum/redhat/7/x86_64/archive/
```
that are part of that major release would move to

```
https://old-releases.repo.saltstack.com/yum/redhat/7/x86_64/archive/
```

If a point release is affected by a CVE that is part of a major supported
branch, that point release will also be moved to
https://old-releases.repo.saltstack.com. For example,

```
https://repo.saltstack.com/yum/redhat/7/x86_64/archive/3000/
```
was affected by the recent CVE so it would be moved to

```
https://old-releases.repo.saltstack.com/yum/redhat/7/x86_64/archive/3000/
```

This would mean someone using the bootstrap script would have to specify `-R
old-releases.repo.saltstack.com` if they wanted to install an

> unsupported/insecure release.  Other methods of installation would also require

Reviewer: This is more work but how about adding a warning message to the bootstrap script, when a vulnerable version is being requested for installation?  Goes along with the point made below, "to make it explicit, and not easy, to install older versions."  This may be wider-reaching than just `old-releases.repo.saltstack.com`, though, i.e. enough to check the version number rather than checking the source of the package.

Reviewer 2: The bootstrap script, in its current incarnation is already bloated and hard to maintain and worse to test.
On the works is a PoC about a python project capable of bootstrapping Salt(single binary using pyinstaller) which doesn't require any system dependencies to bootstrap.
On the new incarnation, we can add such a warning, on the current incarnation, I'm hesitant since the next stable release after Sodium(3001), Magnesium(3002) should be using either the matured PoC or an alternative since a lot will change with Salt moving also to a single binary approach.

> updating the url with the new domain name.

Reviewer: ```suggestion
updating the URL with the new domain name.
```

> https://old-releases.repo.saltstack.com will be s3 syncable just like the main repo.

Reviewer: ```suggestion
https://old-releases.repo.saltstack.com will be S3 syncable just like the main repo.
```

> We will add a notice to the mainpage of repo.saltstack.com explaining this

Reviewer: ```suggestion
We will add a notice to the main page of repo.saltstack.com explaining this
```

behavior.

With regards to the source tarballs hosted on PyPi, we will host them on a web

> server and they will be available on a static HTML page. The reason to do this

Reviewer: ```suggestion
server and they will be made available on a static HTML page. The reason to do this
```
